base/errorshow.jl

```diff
@@ -1104,7 +1104,29 @@ function fielderror_hint_handler(io, exc)
     end
 end
 
-Experimental.register_error_hint(fielderror_hint_handler, FieldError)
+Experimental.register_error_hint(fielderror_dict_hint_handler, FieldError)
+
+function fielderror_listfields_hint_handler(io, exc)
+    fields = fieldnames(exc.type)
+    if isempty(fields)
+        print(io, "; $(nameof(exc.type)) has no fields at all.")
+    else
+        print(io, ", available fields: $(join(map(k -> "`$k`", fields), ", "))")
+    end
+    props = _propertynames_bytype(exc.type)
+    isnothing(props) && return
+    props = setdiff(props, fields)
+    isempty(props) && return
+    print(io, "\nAvailable properties: $(join(map(k -> "`$k`", props), ", "))")
+end
+
+_extract_val(::Type{Val{V}}) where {V} = V
+_extract_val(@nospecialize _) = nothing
+_propertynames_bytype(::Type{T}) where {T} =
+    which(propertynames, (T,)) === which(propertynames, (Any,)) ? nothing :
```

Comment:
```suggestion
```
I think we can simplify `_propertynames_bytype` by removing this branch. If the methods are the same, then `setdiff(props, fields)`, above, will be empty which has the same impact as returning `nothing` here.

Reply:
That change request was for performance reasons, but the short circuit path is much slower than the default path. I don't think it should be accepted. Removing it results in 1.2-2x faster compile time and 10-15x faster runtime.

```julia-repl
x@x:~$ julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.10.4 (2024-06-04)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using Base: promote_op

julia> _extract_val(::Type{Val{V}}) where {V} = V
_extract_val (generic function with 1 method)

julia> _extract_val(@nospecialize _) = nothing
_extract_val (generic function with 2 methods)

julia> _propertynames_bytype(::Type{T}) where {T} =
           # which(propertynames, (T,)) === which(propertynames, (Any,)) ? nothing :
           _extract_val(promote_op(Val∘propertynames, T))
_propertynames_bytype (generic function with 1 method)

julia> @time @eval _propertynames_bytype(Dict)
  0.011929 seconds (25.30 k allocations: 1.731 MiB, 99.02% compilation time)
(:slots, :keys, :vals, :ndel, :count, :age, :idxfloor, :maxprobe)

julia> using Chairmarks

julia> @b Dict _propertynames_bytype
95.714 ns (0.03 allocs: 1.159 bytes)

julia> using LinearAlgebra

julia> @time @eval _propertynames_bytype(LinearAlgebra.QRPivoted)
  0.003214 seconds (15.16 k allocations: 1.025 MiB, 97.22% compilation time)
(:R, :Q, :p, :P)

julia> @b LinearAlgebra.QRPivoted _propertynames_bytype
179.878 ns (0.05 allocs: 1.561 bytes)

julia> 
x@x:~$ julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.10.4 (2024-06-04)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using Base: promote_op

julia> _extract_val(::Type{Val{V}}) where {V} = V
_extract_val (generic function with 1 method)

julia> _extract_val(@nospecialize _) = nothing
_extract_val (generic function with 2 methods)

julia> _propertynames_bytype(::Type{T}) where {T} =
           which(propertynames, (T,)) === which(propertynames, (Any,)) ? nothing :
           _extract_val(promote_op(Val∘propertynames, T))
_propertynames_bytype (generic function with 1 method)

julia> @time @eval _propertynames_bytype(Dict)
  0.014774 seconds (62.30 k allocations: 4.289 MiB, 99.48% compilation time)

julia> @b Dict _propertynames_bytype
1.471 μs (4.37 allocs: 233.263 bytes)

julia> using LinearAlgebra

julia> @time @eval _propertynames_bytype(LinearAlgebra.QRPivoted)
  0.006811 seconds (34.88 k allocations: 2.380 MiB, 98.59% compilation time)
(:R, :Q, :p, :P)

julia> using Chairmarks

julia> @b LinearAlgebra.QRPivoted _propertynames_bytype
1.781 μs (6.50 allocs: 368 bytes)
```